From f9f2a33fdd4d45185ea14a6c38dcf47f18556bbd Mon Sep 17 00:00:00 2001
From: Nikita Maslo <nikitamalco203@gmail.com>
Date: Mon, 4 Aug 2025 10:12:47 +0300
Subject: [PATCH] Add all jethub boards rescue mode

---
 include/configs/jethub.h | 206 ++++++++++++++++++++++++++++++++++++---
 1 file changed, 194 insertions(+), 12 deletions(-)

diff --git a/include/configs/jethub.h b/include/configs/jethub.h
index 699b2c3c899..24df0094583 100644
--- a/include/configs/jethub.h
+++ b/include/configs/jethub.h
@@ -11,25 +11,207 @@

 #if defined(CONFIG_MESON_AXG)
 #define BOOTENV_DEV_RESCUE(devtypeu, devtypel, instance) \
-	"bootcmd_rescue=" \
+	"bootcmd_check_button=" \
 		"if gpio input periphs-banks10; then " \
+		"run bootcmd_rescue;" \
+		"else " \
 		"run bootcmd_mmc0; " \
-		"run bootcmd_usb0;" \
-		"fi;\0"
+		"fi;\0" \
+	\
+	"bootcmd_rescue=" \
+		"echo \"Rescue button pressed, checking USB...\"; " \
+		"usb reset; " \
+		"mmc dev 1; " \
+		"setenv loadaddr 0x06000000; " \
+		"if test -e usb 0:1 rescue.img.gz; then " \
+			"echo \"Found rescue.img.gz, starting eMMC flash...\"; " \
+			"echo \"Flashing rescue.img.gz to eMMC...\"; " \
+			"gpio set periphs-banks27; " \
+				"if load usb 0:1 ${loadaddr} rescue.img.gz; then " \
+					"gpio clear periphs-banks26; " \
+					"setexpr imgsize ${filesize}; " \
+					"gzwrite mmc 1 ${loadaddr} ${imgsize}; " \
+					"echo \"Flashing complete.\"; " \
+					"gpio set periphs-banks26; " \
+					"i=0; " \
+					"while test $i -lt 3; do " \
+						"gpio clear periphs-banks27; " \
+						"sleep 0.3; " \
+						"gpio set periphs-banks27; " \
+						"sleep 0.3; " \
+						"setexpr i $i + 1; " \
+					"done; " \
+					"echo \"Resetting...\"; " \
+					"reset; " \
+					"else " \
+					"echo \"Failed to load rescue.img.gz! Resetting...\"; " \
+					"reset; " \
+				"fi" \
+		"else " \
+			"echo \"rescue.img.gz not found, skipping eMMC flash.\"; " \
+		"fi; " \
+		"echo \"Done. Booting rescue system...\"; " \
+		"run bootcmd_mmc0; \0"
 #elif defined(CONFIG_MESON_G12A)
 #define BOOTENV_DEV_RESCUE(devtypeu, devtypel, instance) \
 	"bootcmd_rescue=" \
-		"if gpio input aobus-banks10; then " \
-		"run bootcmd_mmc1; " \
-		"run bootcmd_mmc0; " \
-		"run bootcmd_usb0;" \
-		"fi;\0"
-#else
+		"echo \"Rescue button pressed, checking USB...\"; " \
+		"usb reset; " \
+		"mmc dev 1; " \
+		"setenv loadaddr 0x07300000; " \
+		"if test -e usb 0:1 rescue.img.gz; then " \
+			"echo \"Found rescue.img.gz, starting eMMC flash...\"; " \
+			"run rescue_flash; " \
+		"else " \
+			"echo \"rescue.img.gz not found, skipping eMMC flash.\"; " \
+		"fi; " \
+		"if test -e usb 0:1 spi-nor.img; then " \
+			"echo \"Found spi-nor.img, checking for SPI update...\"; " \
+			"run update_spi; " \
+		"else " \
+			"echo \"spi-nor.img not found, skipping SPI update.\"; " \
+		"fi; " \
+		"echo \"Done. Booting rescue system...\"; " \
+		"run boot_script; " \
+    "\0" \
+	\
+	"rescue_flash=" \
+    "echo \"Flashing rescue.img.gz to eMMC...\"; " \
+	"gpio set aobus-banks11; " \
+    "if load usb 0:1 ${loadaddr} rescue.img.gz; then " \
+        "setexpr imgsize ${filesize}; " \
+        "gzwrite mmc 1 ${loadaddr} ${imgsize}; " \
+        "echo \"Flashing complete.\"; " \
+		"gpio clear aobus-banks11; " \
+		"i=0; " \
+        "while test $i -lt 3; do " \
+            "gpio set periphs-banks21; " \
+            "sleep 0.3; " \
+            "gpio clear periphs-banks21; " \
+            "sleep 0.3; " \
+            "setexpr i $i + 1; " \
+        "done; " \
+		"echo \"Resetting...\"; " \
+		"reset; " \
+    "else " \
+        "echo \"Failed to load rescue.img.gz! Resetting...\"; " \
+        "reset; " \
+    "fi\0" \
+	\
+	"update_spi=" \
+	"echo \"Checking for spi-nor.img on USB...\"; " \
+	"sf probe; " \
+	"if test -e usb 0:1 spi-nor.img; then " \
+		"echo \"Found spi-nor.img, reading new version...\"; " \
+		"load usb 0:1 ${loadaddr} spi-nor.img 0x40 0x00810000; " \
+		"env import -t ${loadaddr} 0x40; " \
+		"setenv rescue_version_new ${rescue_version}; " \
+		"echo \"New version: ${rescue_version_new}\"; " \
+		"sf read ${loadaddr} 0x00810000 0x40; " \
+		"env import -t ${loadaddr} 0x40; " \
+		"echo \"Current SPI version: ${rescue_version}\"; " \
+		"if itest ${rescue_version_new} -gt ${rescue_version}; then " \
+			"echo \"Newer version found, flashing SPI...\"; " \
+			"gpio set aobus-banks11; " \
+			"load usb 0:1 ${loadaddr} spi-nor.img; " \
+			"setexpr imgsize ${filesize}; " \
+			"sf update ${loadaddr} 0 ${imgsize}; " \
+			"gpio clear aobus-banks11; " \
+			"echo \"SPI update complete. Rebooting...\"; " \
+			"i=0; " \
+			"while test $i -lt 3; do " \
+				"gpio set periphs-banks21; " \
+				"sleep 0.3; " \
+				"gpio clear periphs-banks21; " \
+				"sleep 0.3; " \
+				"setexpr i $i + 1; " \
+			"done; " \
+			"run boot_script; " \
+		"else " \
+			"echo \"SPI is already newest version.\"; " \
+			"run boot_script; " \
+		"fi; " \
+		"else " \
+		"echo \"spi-nor.img not found on USB.\"; " \
+		"run boot_script; " \
+	"fi\0" \
+	\
+	"boot_script=" \
+		"sf probe;" \
+		"setenv bootargs_root \"root=/dev/mtdblock5 ro rootwait rootfs=erofs\"; " \
+		"setenv bootargs_console \"console=tty0 console=ttyAML0,115200n8 fbcon=map:1\"; " \
+		"setenv bootargs_mtdparts \"mtdparts=spi0.0:2M(uboot),64k(env),6M(kernel),64k(version),-(rootfs) mtdids=nor0=spi0.0\"; " \
+		"setenv bootargs \"${bootargs_console} ${bootargs_root} ${bootargs_mtdparts}\"; " \
+		"echo \"Loading LZMA-compressed kernel from spi\"; " \
+		"if sf read 0x10000000 0x00210000 0x00600000; then " \
+			"echo \"Image.lzma loaded successfully.\"; " \
+			"if lzmadec 0x10000000 ${kernel_addr_r}; then " \
+				"echo \"LZMA decompression successful.\"; " \
+			"else " \
+				"echo \"Error: LZMA decompression failed!\"; " \
+				"reset; " \
+			"fi; " \
+		"else " \
+			"echo \"Error: Failed to load Image.lzma from SPI!\"; " \
+			"reset; " \
+		"fi; " \
+		"printenv bootargs;" \
+		"echo \"Starting kernel\"; " \
+		"fdt addr ${fdtcontroladdr}; " \
+		"booti ${kernel_addr_r} - ${fdtcontroladdr}; " \
+		"echo \"Boot failed, resetting...\"; " \
+	"reset\0" \
+	\
+	"bootcmd_check_button=" \
+    "gpio input periphs-banks70; " \
+    "setenv button_state $?; " \
+    "if test ${button_state} -eq 0; then " \
+        "run bootcmd_rescue; " \
+    "else " \
+        "run boot_script; " \
+    "fi\0"
+#elif defined(CONFIG_MESON_GXL)
 #define BOOTENV_DEV_RESCUE(devtypeu, devtypel, instance) \
 	"bootcmd_rescue=" \
-		"if test \"${userbutton}\" = \"true\"; then " \
-		"run bootcmd_mmc0; " \
-		"fi;\0"
+		"echo \"Rescue button pressed, checking USB...\"; " \
+		"usb reset; " \
+		"mmc dev 1; " \
+		"setenv loadaddr 0x10200000; " \
+		"if test -e usb 0:1 rescue.img.gz; then " \
+			"echo \"Found rescue.img.gz, starting eMMC flash...\"; " \
+			"echo \"Flashing rescue.img.gz to eMMC...\"; " \
+			"gpio set periphs-banks73; " \
+				"if load usb 0:1 ${loadaddr} rescue.img.gz; then " \
+					"setexpr imgsize ${filesize}; " \
+					"gzwrite mmc 1 ${loadaddr} ${imgsize}; " \
+					"echo \"Flashing complete.\"; " \
+					"gpio clear periphs-banks73; " \
+					"i=0; " \
+					"while test $i -lt 3; do " \
+						"gpio set periphs-banks73; " \
+						"sleep 0.3; " \
+						"gpio clear periphs-banks73; " \
+						"sleep 0.3; " \
+						"setexpr i $i + 1; " \
+					"done; " \
+					"echo \"Resetting...\"; " \
+					"reset; " \
+					"else " \
+					"echo \"Failed to load rescue.img.gz! Resetting...\"; " \
+					"reset; " \
+				"fi;" \
+			"else" \
+			"echo \"rescue.img.gz not found, skipping eMMC flash.\"; " \
+		"fi; " \
+		"echo \"Done. Booting rescue system...\"; " \
+		"run bootcmd_mmc0; \0" \
+	\
+	"bootcmd_check_button=" \
+	"if test \"${userbutton}\" = \"true\"; then " \
+			"run bootcmd_rescue; " \
+		"else " \
+			"run bootcmd_mmc0; " \
+		"fi\0"
 #endif

 #define BOOTENV_DEV_NAME_RESCUE(devtypeu, devtypel, instance) \
--
2.43.0

