From 09c29a9dfa38745db5abbef9cfe42a9281d2a31e Mon Sep 17 00:00:00 2001
From: Nikita Maslo <nikitamalco203@gmail.com>
Date: Tue, 3 Jun 2025 16:32:04 +0300
Subject: [PATCH] add resque usb script

---
 include/configs/jethub.h | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/include/configs/jethub.h b/include/configs/jethub.h
index 699b2c3c899..9de226b9218 100644
--- a/include/configs/jethub.h
+++ b/include/configs/jethub.h
@@ -28,6 +28,16 @@
 #define BOOTENV_DEV_RESCUE(devtypeu, devtypel, instance) \
 	"bootcmd_rescue=" \
 		"if test \"${userbutton}\" = \"true\"; then " \
+		"echo \"Initializing USB...\"; " \
+		"usb reset; " \
+		"if test -e usb 0:1 rescue.img; then " \
+		"echo \"Rescue image found, starting recovery...\"; " \
+		"load usb 0:1 ${loadaddr} rescue.img; " \
+		"mmc write ${loadaddr} 0x30000 0x4000; " \
+		"reset; " \
+		"else " \
+		"echo \"No rescue image found, booting normally...\"; " \
+		"fi; " \
 		"run bootcmd_mmc0; " \
 		"fi;\0"
 #endif
-- 
2.43.0

