From aa5e13bcf0d33c4a9114ffe9a65460220f2f1868 Mon Sep 17 00:00:00 2001
From: Nikita Maslo <nikitamalco203@gmail.com>
Date: Thu, 7 Aug 2025 11:27:16 +0300
Subject: [PATCH] add spi update uboot script

---
 include/configs/jethub.h | 65 ++++++++++++++++++++++++++++++----------
 1 file changed, 50 insertions(+), 15 deletions(-)

diff --git a/include/configs/jethub.h b/include/configs/jethub.h
index 43c80cf8163..a791aecb6a6 100644
--- a/include/configs/jethub.h
+++ b/include/configs/jethub.h
@@ -19,19 +19,25 @@
 #elif defined(CONFIG_MESON_G12A)
 #define BOOTENV_DEV_RESCUE(devtypeu, devtypel, instance) \
 	"bootcmd_rescue=" \
-	"echo \"Rescue button pressed, checking for rescue.img.gz...\"; " \
-    "usb reset; " \
-    "mmc dev 1; " \
-    "setenv loadaddr 0x48000000; " \
-    "if test -e usb 0:1 rescue.img.gz; then " \
-        "echo \"Found rescue.img.gz, starting flash...\"; " \
-        "run rescue_flash; " \
-		"echo \"Flashing done. Booting from eMMC\"; " \
-        "reset; " \
-    "else " \
-        "echo \"Error: rescue.img.gz not found. Reseting...\"; " \
-        "reset; " \
-    "fi\0" \
+		"echo \"Rescue button pressed, checking USB...\"; " \
+		"usb reset; " \
+		"mmc dev 1; " \
+		"setenv loadaddr 0x48000000; " \
+		"if test -e usb 0:1 rescue.img.gz; then " \
+			"echo \"Found rescue.img.gz, starting eMMC flash...\"; " \
+			"run rescue_flash; " \
+		"else " \
+			"echo \"rescue.img.gz not found, skipping eMMC flash.\"; " \
+		"fi; " \
+		"if test -e usb 0:1 spi-nor.img; then " \
+			"echo \"Found spi-nor.img, checking for SPI update...\"; " \
+			"run update_spi; " \
+		"else " \
+			"echo \"spi-nor.img not found, skipping SPI update.\"; " \
+		"fi; " \
+		"echo \"Done. Booting rescue system...\"; " \
+		"run boot_script; " \
+    "\0" \
 	\
 	"rescue_flash=" \
     "echo \"Flashing rescue.img.gz to eMMC...\"; " \
@@ -39,16 +45,45 @@
         "setexpr imgsize ${filesize}; " \
         "gzwrite mmc 1 ${loadaddr} ${imgsize}; " \
         "echo \"Flashing complete.\"; " \
+		"reset; " \
     "else " \
-        "echo \"Failed to load rescue.img.gz!\"; " \
+        "echo \"Failed to load rescue.img.gz! Resetting...\"; " \
         "reset; " \
     "fi\0" \
+	\
+	"update_spi=" \
+	"echo \"Checking for spi-nor.img on USB...\"; " \
+	"sf probe; " \
+	"if test -e usb 0:1 spi-nor.img; then " \
+		"echo \"Found spi-nor.img, reading new version...\"; " \
+		"load usb 0:1 ${loadaddr} spi-nor.img 0x40 0x00810000; " \
+		"env import -t ${loadaddr} 0x40; " \
+		"setenv rescue_version_new ${rescue_version}; " \
+		"echo \"New version: ${rescue_version_new}\"; " \
+		"sf read ${loadaddr} 0x00810000 0x40; " \
+		"env import -t ${loadaddr} 0x40; " \
+		"echo \"Current SPI version: ${rescue_version}\"; " \
+		"if itest ${rescue_version_new} -gt ${rescue_version}; then " \
+			"echo \"Newer version found, flashing SPI...\"; " \
+			"load usb 0:1 ${loadaddr} spi-nor.img; " \
+			"setexpr imgsize ${filesize}; " \
+			"sf update ${loadaddr} 0 ${imgsize}; " \
+			"echo \"SPI update complete. Rebooting...\"; " \
+			"run boot_script; " \
+		"else " \
+			"echo \"SPI is already newest version.\"; " \
+			"run boot_script; " \
+		"fi; " \
+		"else " \
+		"echo \"spi-nor.img not found on USB.\"; " \
+		"run boot_script; " \
+	"fi\0" \
 	\
 	"boot_script=" \
 		"sf probe;" \
 		"setenv bootargs_root \"root=/dev/mtdblock5 ro rootwait rootfs=erofs\"; " \
 		"setenv bootargs_console \"console=tty0 console=ttyAML0,115200n8 fbcon=map:1\"; " \
-		"setenv bootargs_mtdparts \"mtdparts=spi0.0:2M(uboot),64k(env),6M(kernel),-(rootfs) mtdids=nor0=spi0.0\"; " \
+		"setenv bootargs_mtdparts \"mtdparts=spi0.0:2M(uboot),64k(env),6M(kernel),64k(version),-(rootfs) mtdids=nor0=spi0.0\"; " \
 		"setenv bootargs \"${bootargs_console} ${bootargs_root} ${bootargs_mtdparts}\"; " \
 		"echo \"Loading LZMA-compressed kernel from spi\"; " \
 		"if sf read 0x10000000 0x00210000 0x00600000; then " \
--
2.43.0

