From 25372c5ca87cc481e25e444159f0c6a50978f4cc Mon Sep 17 00:00:00 2001
From: Nikita Maslo <nikitamalco203@gmail.com>
Date: Thu, 7 Aug 2025 11:27:16 +0300
Subject: [PATCH] add spi update uboot script

---
 include/configs/jethub.h | 33 +++++++++++++++++++++++++++++++--
 1 file changed, 31 insertions(+), 2 deletions(-)

diff --git a/include/configs/jethub.h b/include/configs/jethub.h
index 43c80cf8163..eabd884c0e3 100644
--- a/include/configs/jethub.h
+++ b/include/configs/jethub.h
@@ -43,9 +43,38 @@
         "echo \"Failed to load rescue.img.gz!\"; " \
         "reset; " \
     "fi\0" \
+	\
+	"update_spi=" \
+	"echo \"Checking for spi-nor.img on USB...\"; " \
+	"usb reset; " \
+	"sf probe; " \
+	"setenv loadaddr 0x48000000; " \
+	"if test -e usb 0:1 spi-nor.img; then " \
+		"echo \"Found spi-nor.img, reading new version...\"; " \
+		"load usb 0:1 ${loadaddr} spi-nor.img 0x40 0x00800000; " \
+		"env import -t ${loadaddr} 0x40; " \
+		"setenv rescue_version_new ${rescue_version}; " \
+		"echo \"New version: ${rescue_version_new}\"; " \
+		"sf read ${loadaddr} 0x00800000 0x40; " \
+		"env import -t ${loadaddr} 0x40; " \
+		"echo \"Current SPI version: ${rescue_version}\"; " \
+		"if itest ${rescue_version_new} -gt ${rescue_version}; then " \
+			"echo \"Newer version found, flashing SPI...\"; " \
+			"load usb 0:1 ${loadaddr} spi-nor.img; " \
+			"setexpr imgsize ${filesize}; " \
+			"sf update ${loadaddr} 0 ${imgsize};\0"
+			"echo \"SPI update complete. Rebooting...\"; " \
+			"reset; " \
+		"else " \
+			"echo \"SPI is already newest version.\"; " \
+			"run boot_script; " \
+		"fi; " \
+		"else " \
+		"echo \"spi-nor.img not found on USB.\"; " \
+		"run boot_script; " \
+	"fi\0" \
 	\
 	"boot_script=" \
-		"sf probe;" \
 		"setenv bootargs_root \"root=/dev/mtdblock4 ro rootwait rootfs=erofs\"; " \
 		"setenv bootargs_console \"console=tty0 console=ttyAML0,115200n8 fbcon=map:1\"; " \
 		"setenv bootargs_mtdparts \"mtdparts=spi0.0:2M(uboot),64k(env),6M(kernel),-(rootfs) mtdids=nor0=spi0.0\"; " \
@@ -76,7 +105,7 @@
     "if test ${button_state} -eq 0; then " \
         "run bootcmd_rescue; " \
     "else " \
-        "run boot_script; " \
+        "run update_spi; " \
     "fi\0"
 #else
 #define BOOTENV_DEV_RESCUE(devtypeu, devtypel, instance) \
-- 
2.43.0

