From 889a1b0e73d18c769efb62d3a905ccc7043b9be4 Mon Sep 17 00:00:00 2001
From: Nikita Maslo <nikitamalco203@gmail.com>
Date: Thu, 15 May 2025 11:53:04 +0300
Subject: [PATCH] part.h-add-JRESCUE-part-id

---
 include/part.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/part.h b/include/part.h
index fcb3c13dea4..452f98e372d 100644
--- a/include/part.h
+++ b/include/part.h
@@ -32,6 +32,7 @@ struct block_drvr {
 #define PART_TYPE_EFI		0x05
 #define PART_TYPE_MTD		0x06
 #define PART_TYPE_UBI		0x07
+#define PART_TYPE_JRESCUE	0x08
 
 /* maximum number of partition entries supported by search */
 #define DOS_ENTRY_NUMBERS	8
-- 
2.43.0

